native/src/main/java/io/ballerina/stdlib/xmldata/XmlToJson.java

```diff
@@ -118,30 +119,68 @@ public static Object convertToJSON(BXml xml, String attributePrefix, boolean pre
-                                         boolean preserveNamespaces) {
+                                         boolean preserveNamespaces, AttributeManager attributeManager) {
         BMap<BString, Object> rootNode = newJsonMap();
-        LinkedHashMap<String, String> attributeMap = collectAttributesAndNamespaces(xmlItem, preserveNamespaces);
+        BMap<BString, Object> mapData = newJsonMap();
+        if (preserveNamespaces) {
+            processAttributes(xmlItem, preserveNamespaces, attributePrefix, mapData, attributeManager);
+        }
         String keyValue = getElementKey(xmlItem, preserveNamespaces);
-        Object children = convertBXmlSequence((BXmlSequence) xmlItem.getChildrenSeq(),
-                attributePrefix, preserveNamespaces);
-
-        if (children == null) {
-            if (attributeMap.isEmpty()) {
-                putAsBStrings(rootNode, keyValue, "");
+        Object children = convertBXmlSequence((BXmlSequence) xmlItem.getChildrenSeq(), attributePrefix,
+                preserveNamespaces, attributeManager);
+        if (mapData.size() > 0) {
+            if (children == null || children instanceof BString) {
+                if (children != null) {
+                    putAsBStrings(mapData, CONTENT, children.toString().trim());
+                    rootNode = mapData;
+                } else {
+                    putAsBStrings(rootNode, keyValue, mapData);
+                }
             } else {
-                BMap<BString, Object> attrMap = newJsonMap();
-                addAttributes(attrMap, attributePrefix, attributeMap);
-
-                putAsBStrings(rootNode, keyValue, attrMap);
+                BMap<BString, Object> data = (BMap<BString, Object>) children;
+                for (Map.Entry<BString, Object> entry: mapData.entrySet()) {
+                    data.put(entry.getKey(), entry.getValue());
+                }
+                putAsBStrings(rootNode, keyValue, data);
+            }
+        } else {
+            if (children == null) {
+                putAsBStrings(rootNode, keyValue, EMPTY_STRING);
+            } else if (children instanceof BString) {
+                putAsBStrings(rootNode, keyValue, children.toString().trim());
+            } else {
+                putAsBStrings(rootNode, keyValue, children);
             }
-            return rootNode;
         }
-
-        putAsBStrings(rootNode, keyValue, children);
-        addAttributes(rootNode, attributePrefix, attributeMap);
         return rootNode;
     }
 
+    private static void processAttributes(BXmlItem xmlItem, boolean preserveNamespaces, String attributePrefix,
+                                      BMap<BString, Object> mapData, AttributeManager attributeManager) {
+        LinkedHashMap<String, String> tempAttributeMap =  new LinkedHashMap<>();
+        if (attributeManager.getMap().isEmpty()) {
+            tempAttributeMap = collectAttributesAndNamespaces(xmlItem, preserveNamespaces);
```

Comment:
Do we need to call another function with `preserveNamespaces` boolean? Cant we take the decision here?